### DISCUSSION TITLE
How to properly rename a variable from a code-fix

### DISCUSSION THREAD

**Discussion by Evangelink** [dotnet/roslyn#46129] [General]
When trying to implement some analyzer code-fix for rule [RS0046 - Avoid 'opt' suffix](https://github.com/dotnet/roslyn-analyzers/pull/3352) which renames a variable, I have tried to use the `LookupSymbols` API to avoid name clash but I am facing a couple of issues:

### Name clash for method parameters. 

Consider the following method

```csharp
public class C
{
    public void M(string? sOpt, string s, TextSpan? selectionOpt) // Rule correctly flags 'sOpt' and 'selectionOpt' but codefix shall not be offered as they result in name clash
    {
        if (selectionOpt.HasValue)
        {
            var selection = selectionOpt.Value;
            Console.WriteLine(selection);
        }
    }
}
```

but I do not want to register a code-fix which renames `sOpt` into `s` and `selectionOpt` into `selection` because it will cause a name clash. 

Doing some experiments around the `SemanticModel.LookupSymbols` API, I have managed to detect the `sOpt` case by passing a position in the method block (last statement span end) but from this position I cannot detect the variable `selection` which is now out-of-scope. The only thing I can think of is to collect all symbols but that seems rather inefficient and I assume there are lots of analyzers/code-fix in a similar situation.

On https://github.com/dotnet/roslyn-analyzers/issues/3813, @sharwell recommended to have a look at roslyn's name suggestion completion provider, which I guess is https://github.com/dotnet/roslyn/blob/master/src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs but I couldn't really find which part needs to be used.

### Rename all references

I am currently using the `Renamer.RenameSymbolAsync` API to rename the variable but I have noticed that if I rename an interface or base class parameter name, the implementations are not renamed. Is there a way to rename in chain?

Thank you for the help!

**Comment by sharwell** [comment]
The entry point to code to generate a name, accounting for other names in scope, is here:

https://github.com/dotnet/roslyn/blob/41fa23e17b30659610a2b257155b688cd4f5ac9a/src/Features/CSharp/Portable/Completion/CompletionProviders/DeclarationNameCompletionProvider.cs#L250-L257

**Comment by Evangelink** [comment]
@sharwell Thanks for the pointer, regarding `GetLanguageService<ISemanticFactsService>` it seems that we need to reference the workspace package which isn't available in roslyn-analyzers, is it ok or is it going to cause runtime issues?

**Comment by sharwell** [comment]
You'll need to copy all of the implementation necessary for that feature over to roslyn-analyzers.
